add DNS names for TUF repo depot

davepacheco opened the pull request:
I want to be able to use this during blueprint execution for Nexus to fetch SP-related artifacts.

Nexus can obviously just look up the addresses of sleds directly.  But we also want behavior that's already implemented in QorbResolver, which is to have a background task that periodically re-resolves the names and exposes a list of backends via a watch channel.  Providing a DNS name gives us that.

iliana commented on nexus/types/src/deployment/execution/dns.rs:
```diff
@@ -132,6 +132,27 @@ pub fn blueprint_internal_dns_config(
+    // replicated synchronously or atomically to all instances.  That is: a
+    // consumer should be careful when fetching an artifact about whether they
+    // really can just pick any backend of this service or not.
+    for (sled_id, sled) in sleds_by_id {
+        if sled.policy().matches(SledFilter::TufArtifactReplication) {
+            continue;
+        }
```
Isn't this the opposite of what we want?

davepacheco replied:
Ugh, yes -- sorry, I'm still in the process of testing this.

jgallagher commented on nexus/types/src/deployment/execution/utils.rs:
```diff
@@ -8,36 +8,57 @@ use nexus_sled_agent_shared::inventory::SledRole;
 impl Sled {
     pub fn new(
         id: SledUuid,
+        policy: SledPolicy,
```
I think this is fine, but I'm still a little unsettled that blueprint execution operates on the current-as-read-directly-from-the-database `Sled` properties instead of whatever the properties were when the planner ran. Will this work correctly if a sled's policy has changed (in either direction) w.r.t. whether it matches `SledFilter::TufArtifactReplication` since the blueprint was planned?

davepacheco commented:
I tested this by deploying on a4x2 and seeing the names:

```
root@oxz_switch:~# omdb db sleds
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:102::3]:32221,[fd00:1122:3344:101::4]:32221,[fd00:1122:3344:102::4]:32221,[fd00:1122:3344:103::3]:32221,[fd00:1122:3344:101::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (135.0.0)
note: listing all commissioned sleds (use -F to filter, e.g. -F in-service)
 SERIAL  IP                             ROLE      POLICY      STATE   ID                                   
 g3      [fd00:1122:3344:103::1]:12345  scrimlet  in service  active  55bc1b53-7779-48a8-a608-1ec22fd83ad4 
 g0      [fd00:1122:3344:101::1]:12345  scrimlet  in service  active  a7e78a3f-4b18-4280-9b32-29096bccdea1 
 g1      [fd00:1122:3344:102::1]:12345  -         in service  active  fd9cfe11-ec5a-4a0c-a3af-7b4ea0d6d183 

root@oxz_switch:~# dig -t SRV _repo-depot._tcp.control-plane.oxide.internal @fd00:1122:3344:2::1 

; <<>> DiG 9.18.14 <<>> -t SRV _repo-depot._tcp.control-plane.oxide.internal @fd00:1122:3344:2::1
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 60486
;; flags: qr rd; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 3
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;_repo-depot._tcp.control-plane.oxide.internal. IN SRV

;; ANSWER SECTION:
_repo-depot._tcp.control-plane.oxide.internal. 0 IN SRV 0 0 12348 55bc1b53-7779-48a8-a608-1ec22fd83ad4.sled.control-plane.oxide.internal.
_repo-depot._tcp.control-plane.oxide.internal. 0 IN SRV 0 0 12348 a7e78a3f-4b18-4280-9b32-29096bccdea1.sled.control-plane.oxide.internal.
_repo-depot._tcp.control-plane.oxide.internal. 0 IN SRV 0 0 12348 fd9cfe11-ec5a-4a0c-a3af-7b4ea0d6d183.sled.control-plane.oxide.internal.

;; ADDITIONAL SECTION:
55bc1b53-7779-48a8-a608-1ec22fd83ad4.sled.control-plane.oxide.internal. 0 IN AAAA fd00:1122:3344:103::1
a7e78a3f-4b18-4280-9b32-29096bccdea1.sled.control-plane.oxide.internal. 0 IN AAAA fd00:1122:3344:101::1
fd9cfe11-ec5a-4a0c-a3af-7b4ea0d6d183.sled.control-plane.oxide.internal. 0 IN AAAA fd00:1122:3344:102::1

;; Query time: 0 msec
;; SERVER: fd00:1122:3344:2::1#53(fd00:1122:3344:2::1) (UDP)
;; WHEN: Wed Apr 09 19:59:37 UTC 2025
;; MSG SIZE  rcvd: 417

```

and those look like the right addresses and it seems to be a TUF repo depot server (based only on the error message, since I don't have any artifacts here):

```
root@oxz_switch:~# curl http://[fd00:1122:3344:101::1]:12348/artifact/sha256/junk
{
  "request_id": "536d7ad0-9134-4f1d-b007-8aedef7600c5",
  "message": "bad parameter in URL path: Invalid string length"
}
root@oxz_switch:~# curl http://[fd00:1122:3344:101::1]:12348/artifact/sha257/
{
  "request_id": "d375fb9c-3c1e-49d6-ae79-73d953ffd98b",
  "message": "Not Found"
}root@oxz_switch:~# 
```

davepacheco commented:
@jgallagher pointed out that this really only tests the RSS path.  I'd forgotten blueprint execution was not enabled.  So here I enabled the initial blueprint, showed that we successfully executed it, and it didn't change DNS:

```
root@oxz_switch:~# omdb -w nexus blueprints target enable current
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
set target blueprint bf742c7f-8563-4ec5-b372-f4927674d3b8 to enabled
root@oxz_switch:~# omdb nexus background-tasks show blueprint_executor
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
task: "blueprint_executor"
  configured period: every 1m
  currently executing: no
  last completed activation: iter 43, triggered by a dependent task completing
    started at 2025-04-09T20:07:02.745Z (7s ago) and ran for 5114ms
    target blueprint: bf742c7f-8563-4ec5-b372-f4927674d3b8 
    execution:        enabled                              
    status:           completed (13 steps)                 
    error:            (none)                               

root@oxz_switch:~# omdb nexus background-tasks print-report blueprint_executor
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
[Apr 09 20:07:02]   Running ( 1/13) Ensure external networking resources
[Apr 09 20:07:03] Completed ( 1/13) Ensure external networking resources: after 575.82ms
[Apr 09 20:07:03]   Running ( 2/13) Fetch sled list
[Apr 09 20:07:03] Completed ( 2/13) Fetch sled list: after 644.42ms
[Apr 09 20:07:03]   Running ( 3/13) Deploy sled configs
[Apr 09 20:07:05] Completed ( 3/13) Deploy sled configs: after 1.03s
[Apr 09 20:07:05]   Running ( 4/13) Plumb service firewall rules
[Apr 09 20:07:06] Completed ( 4/13) Plumb service firewall rules: after 1.29s
[Apr 09 20:07:06]   Running ( 5/13) Deploy DNS records
[Apr 09 20:07:07] Completed ( 5/13) Deploy DNS records: after 712.72ms
[Apr 09 20:07:07]   Running ( 6/13) Cleanup expunged zones
[Apr 09 20:07:07] Completed ( 6/13) Cleanup expunged zones: after 10.50µs
[Apr 09 20:07:07]   Running ( 7/13) Decommission sleds
[Apr 09 20:07:07] Completed ( 7/13) Decommission sleds: after 5.20µs
[Apr 09 20:07:07]   Running ( 8/13) Decommission expunged disks
[Apr 09 20:07:07] Completed ( 8/13) Decommission expunged disks: after 7.86µs
[Apr 09 20:07:07]   Running ( 9/13) Deploy clickhouse cluster nodes
[Apr 09 20:07:07] Completed ( 9/13) Deploy clickhouse cluster nodes: after 4.53µs
[Apr 09 20:07:07]   Running (10/13) Deploy single-node clickhouse cluster
[Apr 09 20:07:07] Completed (10/13) Deploy single-node clickhouse cluster: after 258.48ms
[Apr 09 20:07:07]   Running (11/13) Mark support bundles as failed if they rely on an expunged disk or sled
[Apr 09 20:07:07] Completed (11/13) Mark support bundles as failed if they rely on an expunged disk or sled: after 526.07ms with message: support bundle expunge report: SupportBundleExpungementReport { bundles_failed_missing_datasets: 0, bundles_deleted_missing_datasets: 0, bundles_failing_missing_nexus: 0, bundles_reassigned: 0 }
[Apr 09 20:07:07]   Running (12/13) Reassign sagas
[Apr 09 20:07:07] Completed (12/13) Reassign sagas: after 12.40ms
[Apr 09 20:07:07]   Running (13/13) Ensure CockroachDB settings
[Apr 09 20:07:07] Completed (13/13) Ensure CockroachDB settings: after 33.65ms
root@oxz_switch:~# omdb db dns show
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:102::3]:32221,[fd00:1122:3344:101::4]:32221,[fd00:1122:3344:102::4]:32221,[fd00:1122:3344:103::3]:32221,[fd00:1122:3344:101::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (135.0.0)
GROUP    ZONE                         ver UPDATED              REASON                  
internal control-plane.oxide.internal 1   2025-04-09T19:26:32Z rack setup              
external oxide.test                   2   2025-04-09T19:26:35Z create silo: "recovery" 
```

davepacheco commented:
I also tested making a different change to DNS.  Here, I added another sled:

```
root@oxz_switch:~# omdb nexus sleds list-uninitialized
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
RACK_ID                              CUBBY SERIAL PART  REVISION 
1cfe1c9b-605b-47cc-ae58-f352e2eaabb1 2     g2     i86pc 0        
root@oxz_switch:~# omdb nexus sleds add g2 i86pc

root@oxz_switch:~# omdb -w nexus sleds add g2 i86pc
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
added sled g2 (i86pc): 0d40e99d-3715-4b92-98f4-4236b9af0ad8

root@oxz_switch:~# omdb -w nexus blueprints regenerate
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
generated new blueprint c783432f-ed9d-43ed-904e-47badb5ed00b

root@oxz_switch:~# omdb -w nexus blueprints diff current c783432f-ed9d-43ed-904e-47badb5ed00b
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
from: blueprint bf742c7f-8563-4ec5-b372-f4927674d3b8
to:   blueprint c783432f-ed9d-43ed-904e-47badb5ed00b

...

 ADDED SLEDS:

  sled 0d40e99d-3715-4b92-98f4-4236b9af0ad8 (active, config generation 2):

    physical disks:
    ---------------------------------------------------------------------------
    vendor             model                serial                  disposition
    ---------------------------------------------------------------------------
+   synthetic-vendor   synthetic-model-U2   synthetic-serial-g2_0   in service 
+   synthetic-vendor   synthetic-model-U2   synthetic-serial-g2_1   in service 
+   synthetic-vendor   synthetic-model-U2   synthetic-serial-g2_2   in service 
+   synthetic-vendor   synthetic-model-U2   synthetic-serial-g2_3   in service 
+   synthetic-vendor   synthetic-model-U2   synthetic-serial-g2_4   in service 


    datasets:
    -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    dataset name                                                                                       dataset id                             disposition   quota     reservation   compression
    -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+   oxp_63fc9e97-7bec-47ed-b61f-d8ae648d3b74/crypt/zone                                                2cd80737-8756-47bb-80d0-c839f6d55a3f   in service    none      none          off        
+   oxp_6715569a-494d-472c-8ef8-5756ea06212f/crypt/zone                                                366e4b60-0e5e-44bd-8a6a-2fab98b7fc9d   in service    none      none          off        
+   oxp_892dd840-3dfb-4ca1-b4ef-f0bbea19a813/crypt/zone                                                cba8233e-1bc5-4695-911c-80363804d9c9   in service    none      none          off        
+   oxp_950144d2-4812-4cec-a6f0-bef014e5d28b/crypt/zone                                                7c02a910-25fa-4861-b4d7-abbed0b91b23   in service    none      none          off        
+   oxp_f71d361f-e252-422a-9a7e-f0d2b09fdda8/crypt/zone                                                18f49f3c-20bd-4cf1-976b-4fc639d141e6   in service    none      none          off        
+   oxp_63fc9e97-7bec-47ed-b61f-d8ae648d3b74/crypt/zone/oxz_ntp_bf04bc38-b67b-4b85-8004-87a9f12e6ce8   8f221c6a-2aea-497e-b0e6-fd8d83bee941   in service    none      none          off        
+   oxp_63fc9e97-7bec-47ed-b61f-d8ae648d3b74/crypt/debug                                               3489de45-3b31-401c-bb61-9f0a6164975b   in service    100 GiB   none          gzip-9     
+   oxp_6715569a-494d-472c-8ef8-5756ea06212f/crypt/debug                                               708db5bc-56fc-4afa-9334-86063d46ddcb   in service    100 GiB   none          gzip-9     
+   oxp_892dd840-3dfb-4ca1-b4ef-f0bbea19a813/crypt/debug                                               67cd7a0c-f0f8-474a-b0fe-5694387faef2   in service    100 GiB   none          gzip-9     
+   oxp_950144d2-4812-4cec-a6f0-bef014e5d28b/crypt/debug                                               bc4a2f9a-5c68-42a6-b95e-718990f8e917   in service    100 GiB   none          gzip-9     
+   oxp_f71d361f-e252-422a-9a7e-f0d2b09fdda8/crypt/debug                                               031e093a-aceb-4fc7-a971-da1ccb59b2f4   in service    100 GiB   none          gzip-9     


    omicron zones:
    ------------------------------------------------------------------------------------------------------------
    zone type      zone id                                image source      disposition   underlay IP           
    ------------------------------------------------------------------------------------------------------------
+   internal_ntp   bf04bc38-b67b-4b85-8004-87a9f12e6ce8   install dataset   in service    fd00:1122:3344:121::21


 COCKROACHDB SETTINGS:
    state fingerprint:::::::::::::::::   d4d87aa2ad877a4cc2fddd0573952362739110de (unchanged)
    cluster.preserve_downgrade_option:   "22.1" (unchanged)

 METADATA:
    internal DNS version:   1 (unchanged)
    external DNS version:   2 (unchanged)

```

```
root@oxz_switch:~# omdb -w nexus blueprints target set c783432f-ed9d-43ed-904e-47badb5ed00b enabled
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
set target blueprint to c783432f-ed9d-43ed-904e-47badb5ed00b
```

Interestingly (but unrelatedly), my first several checks caught a blueprint execution aborted because of the change in blueprint (good to know that works):

```
root@oxz_switch:~# omdb nexus background-tasks show blueprint_executor
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
task: "blueprint_executor"
  configured period: every 1m
  currently executing: iter 48, triggered by a dependent task completing
    started at 2025-04-09T20:10:21.197Z, running for 3938ms
  last completed activation: iter 47, triggered by a periodic timer firing
    started at 2025-04-09T20:10:13.845Z (11s ago) and ran for 7352ms
    target blueprint: bf742c7f-8563-4ec5-b372-f4927674d3b8                                                                                                                                                                         
    execution:        enabled                                                                                                                                                                                                      
    status:           completed (13 steps)                                                                                                                                                                                         
    warning:          at: Mark support bundles as failed if they rely on an expunged disk or sled: Invalid Request: blueprint target has changed from bf742c7f-8563-4ec5-b372-f4927674d3b8 -> c783432f-ed9d-43ed-904e-47badb5ed00b 
    error:            (none)                                                                                                                                                                                                       
```

Eventually we did execute that new blueprint though:

```
root@oxz_switch:~# omdb nexus background-tasks show blueprint_executor
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
task: "blueprint_executor"
  configured period: every 1m
  currently executing: no
  last completed activation: iter 48, triggered by a dependent task completing
    started at 2025-04-09T20:10:21.197Z (35s ago) and ran for 32152ms
    target blueprint: c783432f-ed9d-43ed-904e-47badb5ed00b 
    execution:        enabled                              
    status:           completed (13 steps)                 
    error:            (none)                               

root@oxz_switch:~# omdb nexus background-tasks print-report blueprint_executor
note: Nexus URL not specified.  Will pick one from DNS.
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using Nexus URL http://[fd00:1122:3344:101::6]:12221
[Apr 09 20:10:21]   Running ( 1/13) Ensure external networking resources
[Apr 09 20:10:21] Completed ( 1/13) Ensure external networking resources: after 693.73ms
[Apr 09 20:10:21]   Running ( 2/13) Fetch sled list
[Apr 09 20:10:22] Completed ( 2/13) Fetch sled list: after 340.40ms
[Apr 09 20:10:22]   Running ( 3/13) Deploy sled configs
[Apr 09 20:10:50] Completed ( 3/13) Deploy sled configs: after 27.90s
[Apr 09 20:10:50]   Running ( 4/13) Plumb service firewall rules
[Apr 09 20:10:50] Completed ( 4/13) Plumb service firewall rules: after 453.15ms
[Apr 09 20:10:50]   Running ( 5/13) Deploy DNS records
[Apr 09 20:10:52] Completed ( 5/13) Deploy DNS records: after 2.36s
[Apr 09 20:10:52]   Running ( 6/13) Cleanup expunged zones
[Apr 09 20:10:52] Completed ( 6/13) Cleanup expunged zones: after 12.93µs
[Apr 09 20:10:52]   Running ( 7/13) Decommission sleds
[Apr 09 20:10:52] Completed ( 7/13) Decommission sleds: after 3.75µs
[Apr 09 20:10:52]   Running ( 8/13) Decommission expunged disks
[Apr 09 20:10:52] Completed ( 8/13) Decommission expunged disks: after 5.55µs
[Apr 09 20:10:52]   Running ( 9/13) Deploy clickhouse cluster nodes
[Apr 09 20:10:52] Completed ( 9/13) Deploy clickhouse cluster nodes: after 2.75µs
[Apr 09 20:10:52]   Running (10/13) Deploy single-node clickhouse cluster
[Apr 09 20:10:53] Completed (10/13) Deploy single-node clickhouse cluster: after 57.66ms
[Apr 09 20:10:53]   Running (11/13) Mark support bundles as failed if they rely on an expunged disk or sled
[Apr 09 20:10:53] Completed (11/13) Mark support bundles as failed if they rely on an expunged disk or sled: after 77.02ms with message: support bundle expunge report: SupportBundleExpungementReport { bundles_failed_missing_datasets: 0, bundles_deleted_missing_datasets: 0, bundles_failing_missing_nexus: 0, bundles_reassigned: 0 }
[Apr 09 20:10:53]   Running (12/13) Reassign sagas
[Apr 09 20:10:53] Completed (12/13) Reassign sagas: after 46.33ms
[Apr 09 20:10:53]   Running (13/13) Ensure CockroachDB settings
[Apr 09 20:10:53] Completed (13/13) Ensure CockroachDB settings: after 220.67ms
```

Internal DNS has been updated twice, both in expected ways:

```
root@oxz_switch:~# omdb db dns show
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:102::3]:32221,[fd00:1122:3344:101::4]:32221,[fd00:1122:3344:102::4]:32221,[fd00:1122:3344:103::3]:32221,[fd00:1122:3344:101::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (135.0.0)
GROUP    ZONE                         ver UPDATED              REASON                                                                                                                                                                                                     
internal control-plane.oxide.internal 3   2025-04-09T20:10:52Z blueprint c783432f-ed9d-43ed-904e-47badb5ed00b (sled 0d40e99d-3715-4b92-98f4-4236b9af0ad8: added 5 disks, updated 0, removed 0 disks, sled 0d40e99d-3715-4b92-98f4-4236b9af0ad8: added zone: boundary_ntp) 
external oxide.test                   2   2025-04-09T19:26:35Z create silo: "recovery"                                                                                                                                                                                    
root@oxz_switch:~# omdb db dns diff internal 2
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:102::3]:32221,[fd00:1122:3344:101::4]:32221,[fd00:1122:3344:102::4]:32221,[fd00:1122:3344:103::3]:32221,[fd00:1122:3344:101::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (135.0.0)
DNS zone:                   control-plane.oxide.internal (Internal)
requested version:          2 (created at 2025-04-09T20:10:20Z)
version created by Nexus:   b29e7fb9-ae2e-41f5-82df-71b1b937b945
version created because:    blueprint bf742c7f-8563-4ec5-b372-f4927674d3b8 (initial blueprint from rack setup)
changes:                    names added: 2, names removed: 1

+  0d40e99d-3715-4b92-98f4-4236b9af0ad8.sled          AAAA fd00:1122:3344:121::1
+  _repo-depot._tcp                                   (records: 4)
+      SRV  port 12348 0d40e99d-3715-4b92-98f4-4236b9af0ad8.sled.control-plane.oxide.internal
+      SRV  port 12348 55bc1b53-7779-48a8-a608-1ec22fd83ad4.sled.control-plane.oxide.internal
+      SRV  port 12348 a7e78a3f-4b18-4280-9b32-29096bccdea1.sled.control-plane.oxide.internal
+      SRV  port 12348 fd9cfe11-ec5a-4a0c-a3af-7b4ea0d6d183.sled.control-plane.oxide.internal
-  _repo-depot._tcp                                   (records: 3)
-      SRV  port 12348 55bc1b53-7779-48a8-a608-1ec22fd83ad4.sled.control-plane.oxide.internal
-      SRV  port 12348 a7e78a3f-4b18-4280-9b32-29096bccdea1.sled.control-plane.oxide.internal
-      SRV  port 12348 fd9cfe11-ec5a-4a0c-a3af-7b4ea0d6d183.sled.control-plane.oxide.internal
root@oxz_switch:~# omdb db dns diff internal 3
note: database URL not specified.  Will search DNS.
note: (override with --db-url or OMDB_DB_URL)
note: using DNS server for subnet fd00:1122:3344::/48
note: (if this is not right, use --dns-server to specify an alternate DNS server)
note: using database URL postgresql://root@[fd00:1122:3344:102::3]:32221,[fd00:1122:3344:101::4]:32221,[fd00:1122:3344:102::4]:32221,[fd00:1122:3344:103::3]:32221,[fd00:1122:3344:101::3]:32221/omicron?sslmode=disable
note: database schema version matches expected (135.0.0)
DNS zone:                   control-plane.oxide.internal (Internal)
requested version:          3 (created at 2025-04-09T20:10:52Z)
version created by Nexus:   b29e7fb9-ae2e-41f5-82df-71b1b937b945
version created because:    blueprint c783432f-ed9d-43ed-904e-47badb5ed00b (sled 0d40e99d-3715-4b92-98f4-4236b9af0ad8: added 5 disks, updated 0, removed 0 disks, sled 0d40e99d-3715-4b92-98f4-4236b9af0ad8: added zone: boundary_ntp)
changes:                    names added: 2, names removed: 1

+  _internal-ntp._tcp                                 (records: 2)
+      SRV  port   123 84fe993b-51d6-4cb4-af76-24201b21bfbb.host.control-plane.oxide.internal
+      SRV  port   123 bf04bc38-b67b-4b85-8004-87a9f12e6ce8.host.control-plane.oxide.internal
+  bf04bc38-b67b-4b85-8004-87a9f12e6ce8.host          AAAA fd00:1122:3344:121::21
-  _internal-ntp._tcp                                 (records: 1)
-      SRV  port   123 84fe993b-51d6-4cb4-af76-24201b21bfbb.host.control-plane.oxide.internal

```

I suspect these changes were split into two DNS updates because they happened during two blueprint executions.  The first one must have been a new execution of the old blueprint, which saw the new sled.  This is the behavior discussed [above](https://github.com/oxidecomputer/omicron/pull/7939#discussion_r2035505125).
